### PR TITLE
Add FourColorGradient Sample

### PR DESCRIPTION
**Pull request by hawkerm**
Port of https://www.shadertoy.com/view/7lc3R4

Shows a bit more complexity from the basic gradient sample, has some interesting translations of functions and patterns. Took me a bit to get that `mix` goes to `Lerp`, but works great!

Pretty much did a directly translation of the ShaderToy code. Only thing I changed was abstracting the color constants and making them 4 colors over 2 from the original sample by default.

Edits:

- Flattened the using/namespace directive of the HelloWorld sample to match the others.
- Added a new 'ShaderToImage' sample which shows how to create a blank texture, render a single frame of a shader to it, and save it to a file (like the example in the readme).
- Used that to generate an output image and added a shader unit test for the new shader as requested (as well as compute version).

### CODE REVIEW

File: samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
+﻿namespace ComputeSharp.SwapChain.Shaders
```
Comment (Sergio0694) [inline samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs:1]: Nit: should be using file-scoped using directives

File: samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        /// </summary>
+        public readonly float time;
+
+        // Colors to blend together
+        private static readonly float3 colorOne = new(0.999f, 0.999f, 0.999f);
+        private static readonly float3 colorTwo = new(0.999f, 0f, 0f);
+        private static readonly float3 colorThree = new(0f, 0.999f, 0f);
+        private static readonly float3 colorFour = new(0f, 0f, 0.999f);
+
+        private static float2x2 Rotate(float a)
```
Comment (Sergio0694) [inline samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs:23]: Nit: full XML comments for all these methods. If they're the same as the ones in other sample shaders you can also just copy-paste them from there.

File: samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+            return Hlsl.Frac(Hlsl.Sin(p) * 43758.5453f);
+        }
+
+        private static float Noise(float2 p)
+        {
+            float2 i = Hlsl.Floor(p);
+            float2 f = Hlsl.Frac(p);
+
+            float2 u = f * f * (3.0f - 2.0f * f);
```
Comment (Sergio0694) [inline samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs:43]: Nit: unnecessary blank line.
```suggestion
            float2 u = f * f * (3.0f - 2.0f * f);
```

File: samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            float n = Hlsl.Lerp(
+                        Hlsl.Lerp(
+                            Hlsl.Dot(-1.0f + 2.0f * Hash(i + new float2(0.0f, 0.0f)), f - new float2(0.0f, 0.0f)),
+                            Hlsl.Dot(-1.0f + 2.0f * Hash(i + new float2(1.0f, 0.0f)), f - new float2(1.0f, 0.0f)),
+                            u.X),
+                        Hlsl.Lerp(
+                            Hlsl.Dot(-1.0f + 2.0f * Hash(i + new float2(0.0f, 1.0f)), f - new float2(0.0f, 1.0f)),
+                            Hlsl.Dot(-1.0f + 2.0f * Hash(i + new float2(1.0f, 1.0f)), f - new float2(1.0f, 1.0f)), 
+                            u.X), 
+                        u.Y);
```
Comment (Sergio0694) [inline samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs:55]: Nit: incorrect spacing.
```suggestion
            float n = Hlsl.Lerp(
                Hlsl.Lerp(
                    Hlsl.Dot(-1.0f + 2.0f * Hash(i + new float2(0.0f, 0.0f)), f - new float2(0.0f, 0.0f)),
                    Hlsl.Dot(-1.0f + 2.0f * Hash(i + new float2(1.0f, 0.0f)), f - new float2(1.0f, 0.0f)),
                    u.X),
                Hlsl.Lerp(
                    Hlsl.Dot(-1.0f + 2.0f * Hash(i + new float2(0.0f, 1.0f)), f - new float2(0.0f, 1.0f)),
                    Hlsl.Dot(-1.0f + 2.0f * Hash(i + new float2(1.0f, 1.0f)), f - new float2(1.0f, 1.0f)), 
                    u.X), 
                u.Y);
```

File: samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+        /// <inheritdoc/>
+        public float4 Execute()
+        {
+            float2 fragCoord = new(ThreadIds.X, DispatchSize.Y - ThreadIds.Y);
+            float2 uv = fragCoord / DispatchSize.XY;
+            float ratio = DispatchSize.X / DispatchSize.Y;
+
+            float2 tuv = uv;
+            tuv -= 0.5f;
```
Comment (Sergio0694) [inline samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs:69]: Nit: incorrect spacing.
```suggestion
            float2 tuv = uv;

            tuv -= 0.5f;
```

File: samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+            // rotate with Noise
```
Comment (Sergio0694) [inline samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs:71]: Nit: incorrect capitalization.
```suggestion
            // Rotate with noise
```

File: samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+            tuv.Y *= 1.0f / ratio;
+            tuv.XY = Hlsl.Mul(tuv.XY, Rotate(Hlsl.Radians((degree - 0.5f) * 720.0f + 180.0f)));
+            tuv.Y *= ratio;
+
+            // Wave warp with sin
+            float frequency = 5.0f;
+            float amplitude = 30.0f;
+            float speed = time * 2.0f;
+            tuv.X += Hlsl.Sin(tuv.Y * frequency + speed) / amplitude;
```
Comment (Sergio0694) [inline samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs:82]: Nit: missing blank line between delcarations and statements.
```suggestion
            float speed = time * 2.0f;

            tuv.X += Hlsl.Sin(tuv.Y * frequency + speed) / amplitude;
```

File: samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+            tuv.Y += Hlsl.Sin(tuv.X * frequency * 1.5f + speed) / (amplitude * 0.5f);
+
+            // draw the image
```
Comment (Sergio0694) [inline samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs:85]: Nit: missing uppercase letter.
```suggestion
            // Draw the image
```

File: samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+            float3 layer1 = Hlsl.Lerp(colorOne, colorTwo, Hlsl.SmoothStep(-0.3f, 0.2f, Hlsl.Mul(tuv.XY, Rotate(Hlsl.Radians(-5.0f))).X));
+            float3 layer2 = Hlsl.Lerp(colorThree, colorFour, Hlsl.SmoothStep(-0.3f, 0.2f, Hlsl.Mul(tuv.XY, Rotate(Hlsl.Radians(-5.0f))).X));
+
+            float3 finalComp = Hlsl.Lerp(layer1, layer2, Hlsl.SmoothStep(0.5f, -0.3f, tuv.Y));
+
+            float3 col = finalComp;
```
Comment (Sergio0694) [inline samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs:91]: Nit: unnecessary blank lines.
```suggestion
            float3 finalComp = Hlsl.Lerp(layer1, layer2, Hlsl.SmoothStep(0.5f, -0.3f, tuv.Y));
            float3 col = finalComp;
```

File: samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+
+            return new float4(col, 1.0f);
```
Comment (Sergio0694) [inline samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs:93]: Nit: use target-typed `new` for returns.
```suggestion
            return new(col, 1.0f);
```

**Review by Sergio0694 — CHANGES_REQUESTED** [review]
Thank you for the PR 😄

Could you also add tests for this like with the other shaders?
For reference, you should add a test like this:

https://github.com/Sergio0694/ComputeSharp/blob/d935bc1a09b8c62c1b8107f02316869073473e9e/tests/ComputeSharp.Tests/ShadersTests.cs#L19-L26

Basically you'll need to:
- Define a copy of this shader as a compute shader.
- Save a 1280x720 screenshot of the shader (run it manually over a 2D texture of that size, then save it as a PNG image using the built-in ComputeSharp APIs). Include this screenshot into the right folder like the others.
- Write a test method like for the other shaders to verify it. Pick a small delta value that is just a tiny bit bigger than the minimum value that makes the test pass on your GPU (I'll double check this one, no worries).

Thanks! 😊

**Comment by hawkerm** [comment]
Ah, cool didn't see you had a way to test the shader, that's neat. Should have instructions about adding shaders to have all these steps :)

File: samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs (hunk 1 above)
Comment (hawkerm) [reply samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs:1]: Ah, I had copied the HelloWorld example which doesn't do that. I'll update that as well.

File: samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs (hunk 2 above)
Comment (hawkerm) [reply samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs]: Not all the other samples have this. Also the Rotation one seems really common, why isn't it part of the GLSL api? Does it make sense to have some built-in helpers for these commonly used shader patterns as extensions or something to the `Hlsl` class?

**Comment by hawkerm** [comment]
Had to embiggen the number a bit to pass the Warp tests, but works!

image

Also added a small sample to help generate images in the future. ;)

File: ComputeSharp.sln [hunk 11]
```diff
@@ -48,6 +48,8 @@ Project("{6EC3EE1D-3C4E-46DD-8F32-0CC8E7565705}") = "ComputeSharp.Sample.FSharp"
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "ComputeSharp.Sample.FSharp.Shaders", "samples\ComputeSharp.Sample.FSharp.Shaders\ComputeSharp.Sample.FSharp.Shaders.csproj", "{1F37A704-3CDF-45BB-8114-ED9137E5F2CF}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "ComputeSharp.ShaderToImage", "samples\ComputeSharp.ShaderToImage\ComputeSharp.ShaderToImage.csproj", "{F0FA942A-076E-43C1-B84B-0294FA8DCD91}"
```
Comment (Sergio0694) [inline ComputeSharp.sln:51]: I wish you had asked me about doing this before doing so 😅
I don't want a project just to generate assets for testing.
Could you please rebase and drop this commit?

File: samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs [hunk 12]
```diff
@@ -0,0 +1,106 @@
+﻿namespace ComputeSharp.SwapChain.Shaders;
```
Comment (Sergio0694) [inline samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs:1]: Moving this to file-scoped namespaces in a separate commit creates issues with git blame. Please either add the commit where you did so in the `.git_ignore_revs` file, or (better) just squash this commit on the one where you added this file in the first place directly.

File: samples/ComputeSharp.SwapChain/Shaders/HelloWorld.cs [hunk 13]
```diff
@@ -1,28 +1,27 @@
-﻿namespace ComputeSharp.SwapChain.Shaders
+﻿namespace ComputeSharp.SwapChain.Shaders;
```
Comment (Sergio0694) [inline samples/ComputeSharp.SwapChain/Shaders/HelloWorld.cs:1]: Please drop this commit as well, as it's unrelated from this PR.
Also this missed the compute shader copy of this shader anyway. Let's just do both in a separate PR 🙂

**Review by Sergio0694 — CHANGES_REQUESTED** [review]
Looks great!
Just some more nits and it'll be good to go! 😄

File: ComputeSharp.sln (hunk 11 above)
Comment (hawkerm) [reply ComputeSharp.sln:51]: I mean I used it for testing, but it's still a practical sample based off of the example in the [readme here](https://github.com/Sergio0694/ComputeSharp#working-with-images). We could have the similar prompt like the SwapChain sample, but figured it'd just be simpler to show it working with a single one for now.

Comment (Sergio0694) [reply ComputeSharp.sln:51]: It's used for testing but it's not a project that actually gets used. It's not an actual test to run in the CI nor a sample to use, it's just a one-liner that you need to run to create the .png while creating a new sample. I don't feel like it's worth adding to the solution, especially not now that I have a bazillion branches open and I'll already have to deal with tons of merge conflicts in the solution file when trying to put everything back together 🙂

Comment (hawkerm) [reply ComputeSharp.sln:51]: Since it's just three lines will add to readme instead.

File: samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+        float c = Hlsl.Cos(a);
+        float s = Hlsl.Sin(a);
+
+        return new(c, s, -s, c);
+    }
+
+    /// <summary>
+    /// Generate a hash from a vector.
+    /// </summary>
+    /// <param name="p">Vector to hash</param>
```
Comment (Sergio0694) [inline samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs:37]: Nit: missing trailing '.' on XML parameters and returns.

**Review by Sergio0694 — CHANGES_REQUESTED** [review]

File: tests/ComputeSharp.Tests/Shaders/Compute/FourColorGradient.cs [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+        tuv.X += Hlsl.Sin(tuv.Y * frequency + speed) / amplitude;
+        tuv.Y += Hlsl.Sin(tuv.X * frequency * 1.5f + speed) / (amplitude * 0.5f);
+
+        // Draw the image
+        float3 layer1 = Hlsl.Lerp(colorOne, colorTwo, Hlsl.SmoothStep(-0.3f, 0.2f, Hlsl.Mul(tuv.XY, Rotate(Hlsl.Radians(-5.0f))).X));
+        float3 layer2 = Hlsl.Lerp(colorThree, colorFour, Hlsl.SmoothStep(-0.3f, 0.2f, Hlsl.Mul(tuv.XY, Rotate(Hlsl.Radians(-5.0f))).X));
+        float3 finalComp = Hlsl.Lerp(layer1, layer2, Hlsl.SmoothStep(0.5f, -0.3f, tuv.Y));
+        float3 col = finalComp;
+
+        texture[ThreadIds.XY] = new(col, 1.0f);
```
Comment (Sergio0694) [inline tests/ComputeSharp.Tests/Shaders/Compute/FourColorGradient.cs:109]: Nit: should have the explicit type here as the target type isn't obvious,
Also that `col` temp isn't really needed since it doesn't do anything.
```suggestion
        float3 col = Hlsl.Lerp(layer1, layer2, Hlsl.SmoothStep(0.5f, -0.3f, tuv.Y));

        texture[ThreadIds.XY] = new float4(col, 1.0f);
```

File: samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+        tuv.X += Hlsl.Sin(tuv.Y * frequency + speed) / amplitude;
+        tuv.Y += Hlsl.Sin(tuv.X * frequency * 1.5f + speed) / (amplitude * 0.5f);
+
+        // Draw the image
+        float3 layer1 = Hlsl.Lerp(colorOne, colorTwo, Hlsl.SmoothStep(-0.3f, 0.2f, Hlsl.Mul(tuv.XY, Rotate(Hlsl.Radians(-5.0f))).X));
+        float3 layer2 = Hlsl.Lerp(colorThree, colorFour, Hlsl.SmoothStep(-0.3f, 0.2f, Hlsl.Mul(tuv.XY, Rotate(Hlsl.Radians(-5.0f))).X));
+        float3 finalComp = Hlsl.Lerp(layer1, layer2, Hlsl.SmoothStep(0.5f, -0.3f, tuv.Y));
+        float3 col = finalComp;
+
+        return new(col, 1.0f);
```
Comment (Sergio0694) [inline samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs:104]: Same here as below, except the target-typed `new` is fine as it's in a return statement.
```suggestion
        float3 col = Hlsl.Lerp(layer1, layer2, Hlsl.SmoothStep(0.5f, -0.3f, tuv.Y));

        return new(col, 1.0f);
```

**Review by Sergio0694 — CHANGES_REQUESTED** [review]

File: samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs (hunk 12 above)
Comment (hawkerm) [reply samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs:1]: Fixed

File: samples/ComputeSharp.SwapChain/Shaders/HelloWorld.cs (hunk 13 above)
Comment (hawkerm) [reply samples/ComputeSharp.SwapChain/Shaders/HelloWorld.cs]: Done

File: tests/ComputeSharp.Tests/Shaders/Compute/FourColorGradient.cs (hunk 15 above)
Comment (hawkerm) [reply tests/ComputeSharp.Tests/Shaders/Compute/FourColorGradient.cs]: Done in 58ed2478ed59c8a63794157b45b84239b9116c7f

File: samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs (hunk 16 above)
Comment (hawkerm) [reply samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs]: Done in 58ed2478ed59c8a63794157b45b84239b9116c7f

File: samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs (hunk 14 above)
Comment (hawkerm) [reply samples/ComputeSharp.SwapChain/Shaders/FourColorGradient.cs]: Fixed in all places in 58ed2478ed59c8a63794157b45b84239b9116c7f

**Review by Sergio0694 — APPROVED** [review]
This is awesome! 🙌
Thank you so much for the patience with all the review comments!
